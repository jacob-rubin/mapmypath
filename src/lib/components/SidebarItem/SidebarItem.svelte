<script lang="ts">
	import { fade } from 'svelte/transition';
	import MdiDelete from '~icons/mdi/delete';
	import MdiLocationRadius from '~icons/mdi/location-radius';
	import type Marker from '$lib/utils/marker/marker.svelte';
	import type { MapState } from '$lib/state/mapState/mapState.svelte';
	import { getMapStateContext } from '../Map/utils/mapStateContext';
	import { flip } from 'svelte/animate';

	interface Props {
		marker: Marker;
	}

	let { marker }: Props = $props();
	let isHovering: boolean = $state(false);
	let mapState: MapState = getMapStateContext();

	function remove() {
		mapState.deleteMarker(marker.id);
	}
</script>

<div
	onmouseenter={() => (isHovering = true)}
	onmouseleave={() => (isHovering = false)}
	role="menuitem"
	tabindex={marker.id}
	data-testid="sidebar-item-{marker.id}"
	class={`relative m-2 flex flex-col rounded p-2 outline outline-offset-2 ${isHovering ? 'outline-4' : 'outline-2'}`}
>
	{#if isHovering}
		<button
			class="btn btn-square btn-sm absolute right-0 top-0"
			onclick={remove}
		>
			<MdiDelete class="size-4" />
		</button>
	{/if}
	<input
		type="text"
		bind:value={marker.name}
		class="input w-full max-w-xs"
	/>
	<div class="flex h-8 items-center gap-1 pt-2">
		<MdiLocationRadius class="size-4 shrink-0" />
		<div data-testid="geocode">
			{#await marker.getGeocodeName() then geocodeName}
				<div class="overflow-none text-xs" transition:fade>
					{geocodeName}
				</div>
			{/await}
		</div>
	</div>
</div>
