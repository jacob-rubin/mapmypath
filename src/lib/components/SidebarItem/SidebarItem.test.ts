import {
	cleanup,
	getByRole,
	queryByRole,
	render,
	waitFor
} from '@testing-library/svelte';
import { afterEach, describe, it } from 'vitest';
import SidebarItem from './SidebarItem.svelte';
import mapboxgl from 'mapbox-gl';
import userEvent, {
	type UserEvent
} from '@testing-library/user-event';
import Marker from '$lib/utils/marker/marker.svelte';

describe('SidebarItem', async () => {
	afterEach(() => {
		cleanup();
	});

	it('matches the snapshot', async ({ expect }) => {
		const screen = render(SidebarItem, {
			marker: new Marker({
				id: 1,
				lngLat: new mapboxgl.LngLat(
					-77.03654979172663,
					38.89763503472804
				)
			})
		});
		expect(screen.container.innerHTML).toMatchSnapshot();
	});

	it('updates the name', async ({ expect }) => {
		const user = userEvent.setup();
		const marker = new Marker({
			id: 1,
			lngLat: new mapboxgl.LngLat(0, 0)
		});

		const screen = render(SidebarItem, {
			marker
		});

		const input = screen.getByRole('textbox');

		await user.clear(input);
		await user.type(input, 'New Location');

		expect(marker.name).toBe('New Location');
	});

	it('Loads the geocoded name', async ({ expect }) => {
		const marker = new Marker({
			id: 1,
			lngLat: new mapboxgl.LngLat(
				-77.03654979172663,
				38.89763503472804
			)
		});

		const screen = render(SidebarItem, {
			marker
		});

		const geocode = screen.getByTestId('geocode');
		await waitFor(() => {
			expect(geocode).toHaveTextContent(
				'1600 Pennsylvania Avenue Northwest'
			);
		});
	});

	it('thickens the border on hover', async ({ expect }) => {
		const user: UserEvent = userEvent.setup();
		const marker = new Marker({
			id: 1,
			lngLat: new mapboxgl.LngLat(
				-77.03654979172663,
				38.89763503472804
			)
		});

		const screen = render(SidebarItem, {
			marker
		});
		const sidebarItem: HTMLElement = screen.getByRole('menuitem');

		expect(sidebarItem).not.toHaveClass('outline-4');
		await user.hover(sidebarItem);
		expect(sidebarItem).toHaveClass('outline-4');
	});

	it('shows the delete button when hovered', async ({ expect }) => {
		const user: UserEvent = userEvent.setup();
		const marker = new Marker({
			id: 1,
			lngLat: new mapboxgl.LngLat(
				-77.03654979172663,
				38.89763503472804
			)
		});

		const screen = render(SidebarItem, {
			marker
		});
		const sidebarItem: HTMLElement = screen.getByRole('menuitem');

		expect(queryByRole(sidebarItem, 'button')).toBeNull();
		await user.hover(sidebarItem);
		expect(getByRole(sidebarItem, 'button')).toBeVisible();
	});
});
