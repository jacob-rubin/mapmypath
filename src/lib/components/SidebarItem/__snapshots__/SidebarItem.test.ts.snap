// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarItem > matches the snapshot 1`] = `"<div class="m-2 flex flex-col rounded border-2 border-black p-2" role="menuitem"><input type="text" class="input w-full max-w-xs"> <div class="flex flex-auto items-center gap-1 pt-2"><svg viewBox="0 0 24 24" width="1.2em" height="1.2em" class="size-4 shrink-0"><path fill="currentColor" d="M12 2c3.31 0 6 2.66 6 5.95C18 12.41 12 19 12 19S6 12.41 6 7.95C6 4.66 8.69 2 12 2m0 4a2 2 0 0 0-2 2a2 2 0 0 0 2 2a2 2 0 0 0 2-2a2 2 0 0 0-2-2m8 13c0 2.21-3.58 4-8 4s-8-1.79-8-4c0-1.29 1.22-2.44 3.11-3.17l.64.91C6.67 17.19 6 17.81 6 18.5c0 1.38 2.69 2.5 6 2.5s6-1.12 6-2.5c0-.69-.67-1.31-1.75-1.76l.64-.91C18.78 16.56 20 17.71 20 19"></path></svg><!----> <div class="p-0 leading-none" data-testid="geocode"><!----></div></div></div>"`;
